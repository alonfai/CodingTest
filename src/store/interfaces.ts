import { Interfaces as API } from 'api';

/**
 * User input to retrieve city weather information
 */
export interface LocationInput {
  /**
   * City name
   */
  cityName: string;
  /**
   * ISO 3166 country code
   */
  countryCode: string;
}

/**
 * Exception data on API fetch request for city weather
 */
export interface ExceptionData extends LocationInput {
  /**
   * error message returned from the server
   */
  message: string;
}

/**
 * City condition interface data in UI
 */
export interface Condition extends Omit<API.Weather, 'icon'> {
  /**
   * URL for the icon of the weather condition
   */
  icon_url: string;
}

/**
 * Weather data for a city in the store
 */
export interface CityData {
  /**
   * Name of city
   */
  name: string;
  /**
   * Current temprature of the ciuty
   */
  temprature: number;
  /**
   * Weather condition
   */
  conditions: Condition[];
  /**
   * rain information
   */
  rain?: API.Rain;
  /**
   * snow information
   */
  snow?: API.Snow;
  /**
   * wind information
   */
  wind?: API.Wind;
}

/**
 * Full data object for a city with information, loading state and an optional error exceptions from API (if occured while fetching the data)
 */
export interface CityItem {
  /**
   * data collection
   */
  data?: CityData;
  /**
   * Does the
   */
  loading: boolean;

  /**
   * Exception data returned from the server
   */
  error?: string;
}

/**
 * City reducer data reprentation with key of "{city_name}_{country_code}" and value is actual api response
 */
export interface CityStore {
  [id: string]: CityItem;
}

/**
 * Reference to the store main data type
 */
export interface ReduxStore {
  /**
   * Reducer with all cities information
   */
  cities: CityStore;
}
