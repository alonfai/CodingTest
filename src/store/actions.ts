import { createAction } from 'redux-starter-kit';
import { Interfaces as API } from 'api';
import { LocationInput, ExceptionData } from './interfaces';

/**
 * Action that dispatch loading of city weather information
 */
export const LOAD_CITY_INFORMATION = createAction<LocationInput>('weather/get');

/**
 * SUCCESS response action when API return with city weather information
 */
export const LOAD_CITY_INFORMATION_SUCCESS = createAction<API.ApiResponse>('weather/get_success');

/**
 * FAILURE response action when API return with an error object on retriving city weather information
 */
export const LOAD_CITY_INFORMATION_FAILURE = createAction<ExceptionData>('weather/get_failure');
