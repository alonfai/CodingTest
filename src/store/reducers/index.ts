import city from './city';

export default {
  cities: city,
};
