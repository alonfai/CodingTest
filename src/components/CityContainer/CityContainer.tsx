import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components';

import { Interfaces, AppDispatch, actions } from 'store';
import { request } from 'api';
import CityRecord from '../CityRecord';

const Container = styled.div`
  display: flex;
  flex-basis: 30%;
  min-width: 400px;
  flex-direction: column;
  margin: 2rem;
  justify-content: space-evenly;
  border: 1px solid transparent;
  border-color: #ddd;
  padding: 1rem;
`;

const Header = styled.h1`
  font-basis: 2rem;
  text-decoration: underline;
`;

export interface ComponentProps {
  /**
   * location for the city
   */
  location: Interfaces.LocationInput;
}

const CityContainer: React.FC<ComponentProps> = ({ location }) => {
  const itemId = `${location.cityName}_${location.countryCode}`.toLowerCase();
  const cityItem: Interfaces.CityItem = useSelector((state: Interfaces.ReduxStore) => state.cities[itemId]);
  const dispatch = useDispatch<AppDispatch>();

  // Fetch the city information from the API
  useEffect(() => {
    dispatch(actions.LOAD_CITY_INFORMATION(location));
    const fetchData = async () => {
      try {
        const { cityName, countryCode } = location;
        const response = await request(cityName, countryCode);
        dispatch(actions.LOAD_CITY_INFORMATION_SUCCESS(response));
      } catch (e) {
        const exceptionData: Interfaces.ExceptionData = {
          ...location,
          message: e,
        };
        dispatch(actions.LOAD_CITY_INFORMATION_FAILURE(exceptionData));
      }
    };
    fetchData();
  }, [location, dispatch]);

  if (!cityItem) {
    return null;
  }

  return (
    <Container>
      <Header title={location.cityName}>{location.cityName}</Header>
      <CityRecord item={cityItem} />
    </Container>
  );
};

export default CityContainer;
