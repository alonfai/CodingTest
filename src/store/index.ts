import { configureStore } from 'redux-starter-kit';
import * as Interfaces from './interfaces';
import * as actions from './actions';
import rootReducer from './reducers';

const store = configureStore<Interfaces.ReduxStore>({
  reducer: rootReducer,
});

// HMR Support
if (process.env.NODE_ENV === 'development' && (module as any).hot) {
  (module as any).hot.accept('./reducers', () => {
    const newRootReducer = require('./reducers').default;
    store.replaceReducer(newRootReducer);
  });
}

/**
 * Get the type of the store dispatch (different from regular useDispatch() hook interface, as inside the store we use "redux-thunk" as the async middleware)
 */
export type AppDispatch = typeof store.dispatch;

export { Interfaces, actions };

export default store;
