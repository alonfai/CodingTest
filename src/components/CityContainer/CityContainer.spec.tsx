import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';

import store, { Interfaces } from 'store';
import CityContainer, { ComponentProps } from './CityContainer';

describe('CityContainer', () => {
  const location: Interfaces.LocationInput = {
    cityName: 'mountain view',
    countryCode: 'au',
  };
  const defaultProps: ComponentProps = {
    location,
  };
  let renderer: RenderResult;

  const setup = (props = defaultProps) => {
    return render(
      <Provider store={store}>
        <CityContainer {...props} />
      </Provider>,
    );
  };
  beforeEach(() => {
    renderer = setup();
  });

  it('should render successfully', () => {
    const { container } = renderer;
    expect(container.firstChild).toMatchSnapshot();
  });
});
