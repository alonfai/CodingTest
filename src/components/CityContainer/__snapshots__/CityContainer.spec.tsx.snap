// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityContainer should render successfully 1`] = `
<div
  class="sc-ifAKCX LFxbM"
>
  <h1
    class="sc-EHOje jPeloi"
    title="mountain view"
  >
    mountain view
  </h1>
  <div
    class="sc-bdVaJa eUlSw"
    data-testid="loading"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
